h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -0,0 +1,161 @@
+     *
+     * @param sharedTreeSubgraph An instance of {@link SharedTreeSubgraph} to convert
+     * @return An instance of {@link TreeProperties} with some attributes possibly empty if suitable. Never null.
+     */
+    static TreeProperties convertSharedTreeSubgraph(final SharedTreeSubgraph sharedTreeSubgraph) {
+        Objects.requireNonNull(sharedTreeSubgraph);
+
+        final TreeProperties stf = new TreeProperties();
+
+        stf.leftChildren = MemoryManager.malloc4(sharedTreeSubgraph.nodesArray.size());
+        stf.rightChildren = MemoryManager.malloc4(sharedTreeSubgraph.nodesArray.size());
+        stf.descriptions = new String[sharedTreeSubgraph.nodesArray.size()];
+
+        // Set root node's children, there is no guarantee the root node will be number 0
+        stf.rightChildren[0] = sharedTreeSubgraph.rootNode.getRightChild() != null ? sharedTreeSubgraph.rootNode.getRightChild().getNodeNumber() : -1;
+        stf.leftChildren[0] = sharedTreeSubgraph.rootNode.getLeftChild() != null ? sharedTreeSubgraph.rootNode.getLeftChild().getNodeNumber() : -1;
+
+        List<SharedTreeNode> nodesToTraverse = new ArrayList<>();
+        nodesToTraverse.add(sharedTreeSubgraph.rootNode);
+        append(stf.rightChildren, stf.leftChildren,
+                stf.descriptions, nodesToTraverse, -1, false);
+
+        return stf;
+    }
+
+    private static void append(final int[] rightChildren, final int[] leftChildren, final String[] nodesDescriptions,
+                               List<SharedTreeNode> nodesToTraverse, int pointer, boolean visitedRoot) {
+        if(nodesToTraverse.isEmpty()) return;
+
+        List<SharedTreeNode> discoveredNodes = new ArrayList<>();
+
+        for (SharedTreeNode node : nodesToTraverse) {
+            pointer++;
+            final SharedTreeNode leftChild = node.getLeftChild();
+            final SharedTreeNode rightChild = node.getRightChild();
+            if(visitedRoot){
+                nodesDescriptions[pointer] = serializeNodeDescription(node);
+            } else {
+                nodesDescriptions[pointer] = "Root node";
+                visitedRoot = true;
+            }
+
+            if (leftChild != null) {
+                discoveredNodes.add(leftChild);
+                leftChildren[pointer] = leftChild.getNodeNumber();
+            } else {
+                leftChildren[pointer] = NO_CHILD;
+            }
+
+            if (rightChild != null) {
+                discoveredNodes.add(rightChild);
+                rightChildren[pointer] = rightChild.getNodeNumber();
+            } else {
+                rightChildren[pointer] = NO_CHILD;
+            }
+        }
+
+        append(rightChildren, leftChildren, nodesDescriptions, discoveredNodes, pointer, true);
+    }
+
+    private static String serializeNodeDescription(final SharedTreeNode node) {
+        final StringBuffer nodeDescriptionBuffer = new StringBuffer();
```

Comment:
Why do you need StringBuffer (synchronized), can you use StringBuilder instead?

Reply:
Yes, of course. Just a mistake. Will be fixed.

Reply:
Done.